apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java

```diff
@@ -134,11 +134,18 @@ public static void onExitServletService(@Advice.Argument(0) ServletRequest servl
         if (scope != null) {
             scope.close();
         }
-        if (thiz instanceof HttpServlet && servletRequest instanceof HttpServletRequest) {
+        if (servletRequest instanceof HttpServletRequest) {
             Transaction currentTransaction = tracer.currentTransaction();
             if (currentTransaction != null) {
                 final HttpServletRequest httpServletRequest = (HttpServletRequest) servletRequest;
-                ServletTransactionHelper.setTransactionNameByServletClass(httpServletRequest.getMethod(), thiz.getClass(), currentTransaction.getName());
+                if (thiz instanceof HttpServlet) {
+                    ServletTransactionHelper.setTransactionNameByServletClass(httpServletRequest.getMethod(), thiz.getClass(), currentTransaction.getName());
+                } else if ("javax.faces.webapp.FacesServlet".equals(thiz.getClass().getName())) {
```

Comment:
Could this be done in the lifecycle instrumentation or by instrumenting `javax.faces.webapp.FacesServlet` directly? I would prefer this class not to have knowledge of specific frameworks. Ideally, that knowledge should only be in the plugins.

Reply:
I would prefer it won't be here too. I didn't do it in a separate advice as it either means no general instrumentation for `javax.servlet.Servlet` or chained instrumentation of code from two advices, where order matters.
However, doing it from the Lifecycle is a good idea! I will do that.